### DISCUSSION TITLE
Implementing ranking questions using sortable

### DISCUSSION THREAD

**Discussion by saudiwin** [surveydown-dev/surveydown#160] [Feature Requests]
### Description

Hi guys - 

I use a lot of rank order questions, i.e. ones where respondents have to drag and drop options into their preferred order.

there is an rstudio-supported package, sortable, that implements these types of questions for shiny:

https://rstudio.github.io/sortable/

I use these quite a bit and will in an upcoming survey, so I'd be happy to work on a PR. I wouldn't say my shiny skills are amazing but I can follow an API reasonably well.

Let me know if y'all would be interested and I'll work on it. Of course I can always implement this on a fork but I wanted to see if there was interest in a PR first.

Cheers - 

Bob

**Comment by jhelvy** [comment]
Hell yeah! This is so awesome! This was a question type I really wanted to but wasn't sure where to look for this kind of HTML widget. Would definitely be happy to review a PR on this.

As a heads up, I think this one's going to be a little tricky. Some HTML widgets don't render well in the survey.qmd file, and usually we have to define them in the server and then use sd_output() in the survey.qmd file to designate where to display the question.

My recommendation for getting started is to first make a very simple one-page survey where you use the widget in the survey.qmd file. Don't necessarily try to store the results or anything, just see if the widget displays or not. If it does, then that's a really good sign. If it doesn't, Then this is a question type that will be limited to being defined in the server.

**Comment by saudiwin** [comment]
Sure thing, I'll go take a whack at it.

**Comment by saudiwin** [comment]
Update:

sortable needed a small hack to work in a Quarto document--width and height needed to be explicitly set in the HTML widget rather than set to NULL by default. However, it otherwise seems fine.

file: [test_sortable.qmd.zip](https://github.com/user-attachments/files/18277522/test_sortable.qmd.zip)

The attached .qmd has the code from the repo to redefine the function and compiles just fine. I also posted an issue so hopefully can get this resolved in their repo.

**Comment by saudiwin** [comment]
It would be easy to fork the sortable in any case and put in the quick fix. Let me know what next steps might be for importing this into surveydown.

**Comment by jhelvy** [comment]
Okay I just looked at the [issue](https://github.com/rstudio/sortable/issues/117) in the sortable repo. I understand the problem now. I'm not sure if the sortable dev team will make these fixes or not because, as they mentioned, sortable was designed for interactive shiny apps, not static quarto docs, so they're probably wondering why you're trying to do this. It also kind of doesn't make sense outside of the surveydown context (e.g. what could you possibly do with a sortable list in a static quarto doc?). 

I think we have a few options. 

1. We could keep pushing to get this updated in the sortable package itself. 
2. We could add the functions you wrote here into surveydown as something of a fork of sortable that is specific to surveydown's use case.
3. We could restrict this feature to only be used from inside the server. 

Option 1 is self explanatory. For option 2,  I think this could potentially work since users shouldn't actually need to use `rank_list()`
directly. Instead, we would be integrating this feature into `sd_question()` with perhaps the `type` being set to `"ranklist"` or something. So this little fix you wrote would just be an internal function, e.g. `rank_list2()`, that users never see. Instead of writing: 

```r
rank_list(
      text = "Notice that dragging causes items to swap",
      labels = c("Fish","Bread","Apples"),
      input_id = "rank_list_swap",
      css_id = "rank_list_swap2",
      orientation='vertical',
      options = sortable_options(swap = TRUE)
)
```

Users would just write something like:

```r
sd_question(
    id = "rank_list",
    type = "ranklist",
    label = "Notice that dragging causes items to swap",
    options = c("Fish","Bread","Apples")
)
```

For option 3, you can use the current sortable functions just fine so long as you use them in the server in the app.R file. Something like this:

app.R:

```r
library(surveydown)
library(sortable)

server <- function(input, output, session) {

  output$ranklist <- shiny::renderUI({
    rank_list(
      text = "Notice that dragging causes items to swap",
      labels = c("Fish","Bread","Apples"),
      input_id = "rank_list_swap",
      css_id = "rank_list_swap2",
      orientation='vertical',
      options = sortable_options(swap = TRUE)
    )
  })

  sd_server()
}

shiny::shinyApp(ui = sd_ui(), server = server)
```

survey.qmd:

````qmd
---
format: html
echo: false
warning: false
---

```{r}
library(surveydown)
```

::: {#page1 .sd-page}

```{r}
sd_output('ranklist')
sd_next()
```

:::

::: {#end .sd-page}

End

:::
````

Here you have to use `sd_output()` to show the rendered widget. I don't like this as much though. And we would still need to integrate this into `sd_question()` and then probably use that in the server anyway because if it's not inside `sd_question()` then the values won't get stored in the resulting data.

So my recommendation is to work on option 2.

**Comment by jhelvy** [comment]
I suggest looking at how the [`"mc_multiple_buttons"` type is defined](https://github.com/surveydown-dev/surveydown/blob/main/R/ui.R#L355-L372). It could be as simple as dropping in a new type into `sd_question()` and then adding another if / else condition with your `rank_list2()` function in there. But there might still be a few challenges. First, I'm not sure what the widget returns. My guess is a sorted list of the options. If so, then it should work okay just like the different multi-option widgets work. It also may or may not need the `shiny::tagAppendChild()` bit, which is just a way for surveydown to know if a question has been "answered" or not. Some widgets already do this internally, but for others we had to add that.

**Comment by saudiwin** [comment]
Mmm OK I expect we will need to add that bit of extra code as what I've learned about sortable is that it's a general framework for moving things, the drag and drop question form is just one format. But in any case shouldn't be too hard.

Regarding the options, for now I'm gonna fork the repo and make the fix as that's easy to do. I'll keep the issue open with sortable's Github; if they don't respond, I'll just submit a PR at some point as it may just be a resources issue. I see little reason why exposing some arguments would be a bad idea (though Rstudio has lots of coding rules etc so who knows).

For your all's package, there will need to be some kind of workaround for CRAN as CRAN won't allow a non-CRAN dependency (some exceptions exist but I doubt a forked repo will qualify). So either 1) you all wait for sortable to update to CRAN with the fix before updating surveydown on CRAN or 2) we include some exception handling in case the package is not present. My experience with CRAN is that the second option is usually fine. 

I don't really like the idea of porting the rank question code directly into surveydown as surveydown will still have a dependency on sortable, and if there are upstream changes in sortable it can break the downstream code in ugly ways. I may have some experience with similar issues in other packages :). 

Hopefully I'll have a PR for you all soon-ish -- this is good timing as I'm currently implementing a survey and will be using this question type.
